pocs/camera/pyro.py

```diff
@@ -0,0 +1,475 @@
+
+        # Force camera proxy to connect by getting the camera uid.
+        # This will trigger the remote object creation & (re)initialise the camera & focuser,
+        # which can take a long time with real hardware.
+        uid = self._proxy.get_uid()
+        if not uid:
+            msg = "Couldn't connect to {} on {}!".format(self.name, uri)
+            warn(msg)
+            self.logger.error(msg)
+            return
+
+        self._connected = True
+        self._serial_number = uid
+        self.model = self._proxy.model
+        self._file_extension = self._proxy.file_extension
+        self._readout_time = self._proxy.readout_time
+        self.filter_type = self._proxy.filter_type
+        self.logger.debug("{} connected".format(self))
+
+    def take_exposure(self,
+                      seconds=1.0 * u.second,
+                      filename=None,
+                      dark=False,
+                      blocking=False,
+                      timeout=None,
+                      *args,
+                      **kwargs):
+        """
+        Take exposure for a given number of seconds and saves to provided filename.
+
+        Args:
+            seconds (u.second, optional): Length of exposure
+            filename (str, optional): Image is saved to this filename
+            dark (bool, optional): Exposure is a dark frame (don't open shutter), default False
+            blocking (bool, optional): If False (default) returns immediately after starting
+                the exposure, if True will block until it completes.
+            timeout (u.second, optional): Length of time beyond the length the exposure to wait
+                for exposures to complete. If not given will wait indefinitely.
+
+        Returns:
+            threading.Event: Event that will be set when exposure is complete
+
+        """
+        assert self.is_connected, self.logger.error("Camera must be connected for take_exposure!")
+
+        assert filename is not None, self.logger.warning("Must pass filename for take_exposure")
+
+        # Want exposure time as a builtin type for Pyro serialisation
+        if isinstance(seconds, u.Quantity):
+            seconds = seconds.to(u.second)
+            seconds = seconds.value
+
+        if timeout is not None:
+            if isinstance(timeout, u.Quantity):
+                timeout = timeout.to(u.second)
+                timeout = timeout.value
+
+        dir_name, base_name = os.path.split(filename)
+        # Make sure dir_name has one and only one trailing slash, otherwise rsync may fail
+        dir_name = dir_name.rstrip('/') + '/'
+
+        # Make sure proxy is in async mode
+        Pyro4.async(self._proxy, async=True)
+
+        # Start the exposure
+        self.logger.debug('Taking {} second exposure on {}: {}'.format(
+            seconds, self.name, base_name))
+        # Remote method call to start the exposure
+        exposure_result = self._proxy.take_exposure(seconds=seconds,
```

Comment:
I get a bit worried about how much this (and the `autofocus`) method repeat what is in the main camera class. It seems like the only real change is to have the `._proxy` (and the setting of the async) above. Is that a mostly correct assessment?  Seems like it should be a bit easier to proxy some of these methods without duplicating code.

Reply:
Most methods are easy to proxy with minimal code, but `take_exposure` and `autofocus` do require a bit more.

The main reason for this is because I'm doing the transfer of files from the remote computer, etc.,  outside of Pyro. This is what the Pyro docs recommend, it's not intended for moving big binary blobs around. Most of the code in the `take_exposure` and `autofocus` methods of both the `Camera` and `CameraServer` classes is related to setting up these file transfers and cleaning up afterwards.

Most of the rest of the code in this file is converting the types of arguments/return values before/after passing them through Pyro. Pyro's default serialiser (serpent) handles Python builtin types automatically so for the sake of simplicity I make sure I'm only passing builtin types around.